### PR TITLE
[032][BE] Creation of ClassController

### PR DESCRIPTION
**Pull request by NiloJr-sun**
## Asana Link

https://app.asana.com/0/1203011106874563/1203087509751307/f

## Pre-requisite

- [ ] Check .env files for api and client directory
- [ ] Setup postgres environment
- [ ] Postman
- [ ] Existing registered users, starting ID:3 up to ID:7

## Commands

- [ ] in the terminal, go to the project directory and cd into api directory - `cd api`
- [ ] run `php artisan migrate`
- [ ] Open postman and paste `http://127.0.0.1:8000/api/course` to fetch the class list. PS: Set the method to **GET**
- [ ] Open postman and paste `http://127.0.0.1:8000/api/course/{id}` to fetch a specific class. PS: Set the method to **GET**
- [ ] Open postman and paste 
`http://127.0.0.1:8000/api/course?course=EE 131&department=Electrical Engineering&user_ids=3,4,5,6` to create a class. PS: Set the method to **POST**
- [ ] Open postman and paste 
`http://127.0.0.1:8000/api/course/8?course=EE 131 - Updated&department=Computer Engineering&user_ids=3,4,5` to update a class. PS: Set the method to **PUT**
- [ ] Open postman and paste `http://127.0.0.1:8000/api/department/{id}` to delete a specific class. PS: Set the method to **DELETE**
## Expected Output
 
- [ ] CRUD feature for class controller

## Screenshots

image
image
image
image
image
image
image
image

### CODE REVIEW

File: api/app/Http/Controllers/CourseController.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        $user_ids = array_filter(explode(',', $request['user_ids']));
+        foreach ($user_ids as $user_id) {
+            CourseUser::create(
+                [
+                    'user_id' => $user_id,
+                    'course_id' => $course_id,
+                ]
+            );
+        }
```
Comment (ejanton-sun) [inline api/app/Http/Controllers/CourseController.php:43]: instead using foreach, you can use upsert or insert
Ref: https://laravel.com/docs/9.x/eloquent#upserts

File: api/app/Http/Controllers/CourseController.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    public function show($id)
+    {
+        $course = Course::findOrFail($id);
+        $course->courseUsers;
+
+        return response()->json(
+            [
+                'course_detail' => $course,
+
```
Comment (ejanton-sun) [inline api/app/Http/Controllers/CourseController.php:60]: remove extra space
```suggestion
```

File: api/app/Http/Controllers/CourseController.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            [
+                'status' => 'Class created!',
+            ]
+        );
+    }
+
+    public function show($id)
+    {
+        $course = Course::findOrFail($id);
+        $course->courseUsers;
```
Comment (ejanton-sun) [inline api/app/Http/Controllers/CourseController.php:55]: use `with()` when retrieving with relationship

File: api/app/Http/Controllers/CourseController.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        $request->validate(
+            [
+                'course' => ['required'],
+                'department' => ['required'],
+                'user_ids' => ['required'],
+            ]
+        );
+
+        $department_id = Department::where('department', $request['department'])->get('id')->first()->id;
```
Comment (ejanton-sun) [inline api/app/Http/Controllers/CourseController.php:75]: you can use `->first()` 
```suggestion
        $department_id = Department::where('department', $request['department'])->first()->pluck('id)
```

File: api/app/Http/Controllers/CourseController.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        $course = Course::findOrFail($id);
+        $course->update(
+            [
+                'course' => $request->course,
+                'department_id' => $department_id,
+            ]
+        );
```
Comment (ejanton-sun) [inline api/app/Http/Controllers/CourseController.php:82]: I don't think you need to update also the department Id, only the `course`

File: api/app/Http/Controllers/CourseController.php [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        foreach ($user_ids as $user_id) {
+            CourseUser::updateOrCreate(
+                [
+                    'user_id' => $user_id,
+                    'course_id' => $course->id,
+                ]
+            );
+        }
```
Comment (ejanton-sun) [inline api/app/Http/Controllers/CourseController.php:104]: Too much foreach, refactor this one

**Review by ejanton-sun — CHANGES_REQUESTED** [review]

**Review by jasonchuacodes — COMMENTED** [review]
- [ ] need to assign an instructor before course creation

**Pull request merged**